includes/internal/emails/class-email-sender.php

```diff
@@ -0,0 +1,276 @@
+ *
+ * @package sensei
+ */
+
+namespace Sensei\Internal\Emails;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+
+require plugin_dir_path( __DIR__ ) . '../../vendor/autoload.php';
+
+use \Pelago\Emogrifier\CssInliner;
+
+/**
+ * Class Sensei_Email_Sender
+ *
+ * @package Sensei\Internal\Emails
+ */
+class Email_Sender {
+
+	/**
+	 * Email unique identifier meta key.
+	 */
+	public const EMAIL_ID_META_KEY = '_sensei_email_name';
+
+	/**
+	 * Instance of class.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * The constructor.
+	 *
+	 * @var array
+	 */
+	private function __construct() {}
+
+	/**
+	 * Fetches an instance of the class.
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Adds all filters and actions.
+	 */
+	public function init() {
+		/**
+		 * Send email of predefined types with provided contents.
+		 *
+		 * @param string $email_name   The name of the email template.
+		 * @param array  $replacements The placeholder replacements.
+		 */
+		add_action( 'sensei_send_html_email', [ $this, 'send_email' ], 10, 2 );
+	}
+
+	/**
+	 * Send email of type.
+	 *
+	 * @param string $email_name The email type.
+	 * @param array  $replacements The placeholder replacements.
+	 *
+	 * @access private
+	 */
+	public function send_email( $email_name, $replacements ) {
+		$email_post = $this->get_email_post_by_name( $email_name );
+
+		if ( ! $email_post ) {
+			return;
+		}
+
+		/**
+		 * Filter the email replacements.
+		 *
+		 * @since $$next-version$$
+		 * @hook sensei_email_replacements
+		 *
+		 * @param array   $replacements The email replacements.
+		 * @param string  $email_name   The email name.
+		 * @param WP_Post $email_post   The email post.
+		 * @param self    $email_sender The email sender class instance.
+		 *
+		 * @return array The email replacements.
+		 */
+		$replacements = apply_filters( 'sensei_email_replacements', $replacements, $email_name, $email_post, $this );
+
+		$templated_output = $this->get_templated_post_content( $email_post );
+
+		/**
+		 * Filter the email styles.
+		 *
+		 * @since $$next-version$$
+		 * @hook sensei_email_styles
+		 *
+		 * @param string  $style_string The email styles.
+		 * @param string  $email_name   The email name.
+		 * @param WP_Post $email_post   The email post.
+		 * @param self    $email_sender The email sender class instance.
+		 *
+		 * @return string
+		 */
+		$style_string = apply_filters( 'sensei_email_styles', $this->get_header_styles(), $email_name, $email_post, $this );
+
+		$html_output_with_inlined_css = CssInliner::fromHtml( $templated_output )->inlineCss( $style_string )->render();
+
+		$subject_text = wp_strip_all_tags( $email_post->post_title );
+
+		$headers = [
+			'Content-Type: text/html; charset=UTF-8',
+		];
+
+		foreach ( $replacements as $recipient => $replacement ) {
```

Comment:
Hum, I think Email_Generator might suit better for this logic...

Reply:
I think we can keep it in sender to keep the processing together and handle only the data preparation on event in generator. But we can think again if any use case arises later